### PR TITLE
Add in the ability to do freeform selection via ALT+left mouse drawing

### PR DESCRIPTION
**Pull request by CrucisGamma**
### Description of the Change

Add in a way for a user to be able to run a freeform selection in the main graph view (similar to the box select)

### Alternate Designs

The alternate design was modelled on graphics tools

### Why Should This Be In Core?

It's adding to the functionality of the program and not having to do multiple little box selects at one time

### Benefits

The user will be able to do freeform selections to selectively pick out bits and pieces of the graph 

### Possible Drawbacks

Potential for some duplicated code requiring a refactor.  No speed or memory increase was noticed in the testing 

### Verification Process

Ran through multiple selections and box selections over multiple graphs with no loss of speed or noticeable slowdown after some time doing this.  No exceptions noted or odd behaviour encountered.

### Applicable Issues

#1268

### CODE REVIEW

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+        objectFrameMatrix.getRotationMatrix(rotationMatrixt);
+        final Matrix44f rotationMatrixti = new Matrix44f();
+        rotationMatrixti.invert(rotationMatrixt);
+        final Matrix33f rotationMatrix = new Matrix33f();
+        rotationMatrixti.getRotationMatrix(rotationMatrix);
+
+        // Do the vertex positions need mixing?
+        boolean requiresMix = x2Attr != Graph.NOT_FOUND && y2Attr != Graph.NOT_FOUND && z2Attr != Graph.NOT_FOUND;
+        boolean requiresVertexVisibility = vxVisibilityAttr != Graph.NOT_FOUND;
+        boolean requiresTransactionVisibility = txVisibilityAttr != Graph.NOT_FOUND;
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:148]: final. Also could be moved closer to where they are used

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+                final float bottomMostPoint = (sceneLocation.getY() - r) / -sceneLocation.getZ();
+                final float topMostPoint = (sceneLocation.getY() + r) / -sceneLocation.getZ();
+
+                final boolean vertexLeftOfFreeform = rightMostPoint < left;
+                final boolean vertexRightOfFreeform = right < leftMostPoint;
+                final boolean vertexBelowFreeform = topMostPoint < bottom;
+                final boolean vertexAboveFreeform = top < bottomMostPoint;
+
+                if (!vertexLeftOfFreeform && !vertexRightOfFreeform && !vertexBelowFreeform && !vertexAboveFreeform) {
+                    if (inFreeformPolygons(leftMostPoint, topMostPoint)) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:207]: can merge these 2 if statements together (with `&&`)

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+                        }
+                    }
+                }
+
+                if (txSelectedAttr != Graph.NOT_FOUND) {
+                    for (int position = 0; position < txCount; position++) {
+                        final int txId = graph.getTransaction(position);
+                        if (vxIncluded.get(graph.getTransactionSourceVertex(txId)) && vxIncluded.get(graph.getTransactionDestinationVertex(txId)) && !graph.getBooleanValue(txSelectedAttr, txId)) {
+                            if (requiresTransactionVisibility) {
+                                float visibility = graph.getFloatValue(txVisibilityAttr, txId);
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:321]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+                        selectVerticesOperation.setValue(vertex, !graph.getBooleanValue(vxSelectedAttr, vertex));
+                    }
+                }
+
+                if (txSelectedAttr != Graph.NOT_FOUND) {
+                    for (int position = 0; position < txCount; position++) {
+                        final int txId = graph.getTransaction(position);
+                        if (vxIncluded.get(graph.getTransactionSourceVertex(txId)) && vxIncluded.get(graph.getTransactionDestinationVertex(txId))) {
+                            if (requiresTransactionVisibility) {
+                                float visibility = graph.getFloatValue(txVisibilityAttr, txId);
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:344]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java [hunk 5]
```diff
@@ -1101,6 +1139,50 @@ private void performBoxSelection(final GraphReadMethods rg, final Point selectTo
...
+     * @param selectTo the point where selection ends
+     * @param selectFrom the point where selection begins.
+     * @param appendSelection whether or not the selection will be appended to the current selection
+     * @param toggleSelection whether or not the selection will toggle the current selection. Note that if
+     * appendSelection is true, this parameter has no effect.
+     */
+    private void performFreeformSelection(final GraphReadMethods rg, final Point selectTo, final Point selectFrom, final boolean appendSelection, final boolean toggleSelection) {
+
+        // Sort the press/release coordinates to look like a drag from top-left to bottom-right: pressed<=released.
+        Point bottomRight = new Point();
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java:1162]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java [hunk 6]
```diff
@@ -1101,6 +1139,50 @@ private void performBoxSelection(final GraphReadMethods rg, final Point selectTo
...
+            bottomRight.y = selectFrom.y;
+            selectFrom.y = selectTo.y;
+        } else {
+            bottomRight.y = selectTo.y;
+        }
+
+        final float[] boxCameraCoordinates = visualInteraction.windowBoxToCameraBox(freeformModel.getLeftMost(), freeformModel.getRightMost(), freeformModel.getTopMost(), freeformModel.getBottomMost());
+
+        freeformModel.setWindowBoxToCameraBox(boxCameraCoordinates);
+         Float [] transformedVertices = freeformModel.getTransformedVertices();
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java:1180]: final. Also if you could remove the space between `Float` and `[]` that would be great

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java [hunk 7]
```diff
@@ -1101,6 +1139,50 @@ private void performBoxSelection(final GraphReadMethods rg, final Point selectTo
...
+
+        Plugin plugin = new FreeformSelectionPlugin(appendSelection, toggleSelection, VisualGraphUtilities.getCamera(rg), boxCameraCoordinates, transformedVertices, freeformModel.getNumPoints());
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java:1182]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/InteractiveGLVisualProcessor.java [hunk 8]
```diff
@@ -201,6 +206,13 @@ public VisualOperation setSelectionBoxModel(SelectionBoxModel model) {
                 .withId(selectionBoxUpdateId).build());
     }
 
+    @Override
+    public VisualOperation setSelectionFreeformModel(SelectionFreeformModel model) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/InteractiveGLVisualProcessor.java:210]: parameter final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        points.add(endPoint);
+    }
+
+    /**
+     * Add a new point to an existing model
+     *
+     * @param nextPoint The next point the user wants to add to the start
+     * corner).
+     */
+    static private final int THRESHOLD = 8;
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:70]: `static private` -> `private static`

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+    /**
+     * Get the starting corner point of the selection freeform described by this
+     * model
+     *
+     * @return A 3D vector representing the start of the selection freeform, or
+     * <code>null</code> if the model is clear (represents no selection
+     * freeform).
+     */
+    public Point getPoint(int i) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:173]: parameter final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+    public int getBottomMost() {
+        return bottomMost;
+    }
+
+    public Float[] getTransformedVertices() {
+        int width = rightMost - leftMost;
+        int height = bottomMost - topMost;
+        float newWidth = rightBoxCameraCoordinates - leftBoxCameraCoordinates;
+        float newHeight = bottomBoxCameraCoordinates - topBoxCameraCoordinates;
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:221]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+        ArrayList<Float> vertices = new ArrayList<>();
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:223]: final. Also change declaration to `List<Float>` (can keep instatiation as is)

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+        for (int i = 0; i < MAX_POINTS; i++) {
+            float x = (getPoint(i).x - leftMost) / ((float) width) * newWidth + leftBoxCameraCoordinates;
+            float y = (getPoint(i).y - topMost) / ((float) height) * newHeight + topBoxCameraCoordinates;
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:227]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+            vertices.add(x);
+            vertices.add(y);
+        }
+
+        return vertices.toArray(new Float[0]);
+    }
+
+    public void setWindowBoxToCameraBox(float[] boxCameraCoordinates) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformModel.java:236]: parameter final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        final GL3 gl = drawable.getGL().getGL3();
+        String vs = null;
+        String gs = null;
+        String fs = null;
+
+        try {
+            vs = GLTools.loadFile(GLVisualProcessor.class, "shaders/PassThru.vs");
+            gs = GLTools.loadFile(GLVisualProcessor.class, "shaders/PassThruTriangle.gs");
+            fs = GLTools.loadFile(GLVisualProcessor.class, "shaders/PassThru.fs");
+        } catch (IOException ex) {
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java:87]: exception final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+            Logger.getLogger(SelectionFreeformRenderable.class.getName()).log(Level.SEVERE, null, ex);
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java:88]: I would extract the logger into a private static final field (see other classes for examples). This allows the logger to easily be used elsewhere in the class

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+            // Map the vertex buffer.
+            gl.glBindBuffer(GL3.GL_ARRAY_BUFFER, batch.getBufferName(vertexTarget));
+            final ByteBuffer bbuf = gl.glMapBuffer(GL3.GL_ARRAY_BUFFER, GL3.GL_WRITE_ONLY);
+            final FloatBuffer fbuf = bbuf.asFloatBuffer();
+
+            // We should have the same buffer size we started with.
+            assert fbuf.limit() == NUMBER_OF_VERTICES * 3;
+
+            // Find the location of the Freeform in projected coordinates and update the TRIANGLE_FAN coordinates in the vertex buffer.
+            float[] v = new float[150];
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java:146]: final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+            fbuf.put(v);
+
+            gl.glUnmapBuffer(GL3.GL_ARRAY_BUFFER);
+
+            // Disable depth so the freeform shape is drawn over everything else.
+            gl.glDisable(GL3.GL_DEPTH_TEST);
+            gl.glDepthMask(false);
+
+            Matrix44f mvpMatrix = new Matrix44f();
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/renderables/SelectionFreeformRenderable.java:162]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java [hunk 19]
```diff
@@ -0,0 +1,477 @@
...
+            boolean belowHighY = vertY_j > testY;
+            boolean withinYsEdges = belowLowY != belowHighY;
+
+            if (withinYsEdges) {
+                float slopeOfLine = (vertX_j - vertX_i) / (vertY_j - vertY_i);
+                float pointOnLine = (slopeOfLine * (testY - vertY_i)) + vertX_i;
+                boolean isLeftToLine = testX < pointOnLine;
+
+                if (isLeftToLine) {
+                    locatedInPolygon = !locatedInPolygon;
```
[the remark this answers is not in the document]
Comment (CrucisGamma) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:93]: locatedInPolygon can change many times in the for loop.  This is a ray tracing algorithm where you test if a line  crosses a line in a polygon % 2 times.

**Review by arcturus2 — CHANGES_REQUESTED** [review]
@CrucisGamma it works really well and in terms of performance I couldn't see any issues.

Can you please add some doco regarding this and also an entry in `whatsnew.txt` for this please?

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (antares1470) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/select/FreeformSelectionPlugin.java:86]: Right I missed that. As you were :)

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**